Tools/TestWebKitAPI/Tests/WebKitCocoa/Download.mm

```diff
@@ -2278,6 +2323,92 @@ HTTPServer server({
         DownloadCallback::DidFinish,
     });
 }
+
+TEST(WKDownload, PlaceholderPolicyEnable)
+{
+    HTTPServer server({
+        { "/"_s, { 404, { }, "http body"_s } }
+    });
+    NSURL *expectedDownloadFile = tempFileThatDoesNotExist();
+    auto delegate = adoptNS([TestDownloadDelegate new]);
+    auto webView = adoptNS([WKWebView new]);
+    [webView setNavigationDelegate:delegate.get()];
+
+    __block bool didFinish = false;
+    [webView startDownloadUsingRequest:server.request() completionHandler:^(WKDownload *download) {
+        download.delegate = delegate.get();
+        delegate.get().decideDestinationUsingResponse = ^(WKDownload *, NSURLResponse *, NSString *, void (^completionHandler)(NSURL *)) {
+            completionHandler(expectedDownloadFile);
+        };
+        delegate.get().decidePlaceholderPolicy = ^(WKDownload *, void (^completionHandler)(_WKPlaceholderPolicy, NSURL *)) {
+            completionHandler(_WKPlaceholderPolicyEnable, nil);
+        };
+        delegate.get().downloadDidFinish = ^(WKDownload *download) {
+            didFinish = true;
+        };
+    }];
+    Util::run(&didFinish);
+
+    checkFileContents(expectedDownloadFile, "http body"_s);
+
+    if (modernDownloadProgressEnabled()) {
+        checkCallbackRecord(delegate.get(), {
+            DownloadCallback::DecideDestination,
+            DownloadCallback::DecidePlaceholderPolicy,
+            DownloadCallback::DidReceivePlaceholderURL,
+            DownloadCallback::DidReceiveFinalURL,
+            DownloadCallback::DidFinish,
+        });
+    } else {
+        checkCallbackRecord(delegate.get(), {
+            DownloadCallback::DecideDestination,
+            DownloadCallback::DecidePlaceholderPolicy,
+            DownloadCallback::DidFinish,
+        });
+    }
+}
+
+
+TEST(WKDownload, PlaceholderPolicyDisable)
+{
+    HTTPServer server({
+        { "/"_s, { 404, { }, "http body"_s } }
+    });
+    NSURL *expectedDownloadFile = tempFileThatDoesNotExist();
+    auto delegate = adoptNS([TestDownloadDelegate new]);
+    auto webView = adoptNS([WKWebView new]);
+    [webView setNavigationDelegate:delegate.get()];
+
+    auto progressObserver = adoptNS([[ProgressObserver alloc] init]);
+    auto progress = adoptNS([[NSProgress alloc] init]);
```

Comment:
I think we want to observe the progress of the placeholder, since this is where progress is published in this case. In the latest patch, I set the fileURL property to the placeholder URL, in order to hook this up.